apis/apps/v1alpha1/workloadspread_types.go

```diff
@@ -0,0 +1,210 @@
+type AdaptiveWorkloadSpreadStrategy struct {
+	// DisableSimulationSchedule indicates whether to disable the feature of simulation schedule.
+	// Default is false.
+	// Webhook can take a simple general predicates to check whether Pod can be scheduled into this subset,
+	// but it just considers the Node resource and cannot replace scheduler to do richer predicates practically.
+	// +optional
+	DisableSimulationSchedule bool `json:"disableSimulationSchedule,omitempty"`
+
+	// RescheduleCriticalSeconds indicates how long controller will reschedule a schedule failed Pod to the subset that has
+	// redundant capacity after the subset where the Pod lives. If a Pod was scheduled failed and still in a pending status
+	// over RescheduleCriticalSeconds duration, the controller will reschedule it to the suitable subset.
+	// +optional
+	RescheduleCriticalSeconds *int32 `json:"rescheduleCriticalSeconds,omitempty"`
+}
+
+// WorkloadSpreadSubset defines the details of a subset.
+type WorkloadSpreadSubset struct {
+	// Name should be unique between all of the subsets under one WorkloadSpread.
+	Name string `json:"name"`
+
+	// Indicates the node required selector to form the subset.
+	// +optional
+	RequiredNodeSelectorTerm *corev1.NodeSelectorTerm `json:"requiredNodeSelectorTerm,omitempty"`
+
+	// Indicates the node preferred selector to form the subset.
+	// +optional
+	PreferredNodeSelectorTerms []corev1.PreferredSchedulingTerm `json:"preferredNodeSelectorTerms,omitempty"`
+
+	// Indicates the tolerations the pods under this subset have.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// MaxReplicas indicates the desired max replicas of this subset.
+	// +optional
+	MaxReplicas *intstr.IntOrString `json:"maxReplicas,omitempty"`
+
+	// Patch indicates the patch metadata to the Pod.
+	// +optional
+	Patch runtime.RawExtension `json:"patch,omitempty"`
+}
+
+// WorkloadSpreadStatus defines the observed state of WorkloadSpread.
+type WorkloadSpreadStatus struct {
+	// ObservedGeneration is the most recent generation observed for this WorkloadSpread. It corresponds to the
+	// WorkloadSpread's generation, which is updated on mutation by the API Server.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// ObservedWorkloadGeneration is the most recent generation observed for target Workload.
+	// +optional
+	ObservedWorkloadGeneration int64 `json:"observedWorkloadGeneration,omitempty"`
+
+	// Contains the status of each subset. Each element in this array represents one subset
+	// +optional
+	SubsetStatuses []WorkloadSpreadSubsetStatus `json:"subsetStatuses,omitempty"`
+}
+
+// SubsetUnscheduledStatus contains the details for the unscheduled subset.
+type SubsetUnscheduledStatus struct {
+	// Unschedulable is true indicates this Subset cannot be scheduled. The default is false.
+	// +optional
+	Unschedulable bool `json:"unschedulable,omitempty"`
```

Comment:
It is not clear how to define a subset is schedulable or not? Will a single Pod scheduling failure triggers this status update? What if out of x replicas, only one Pod fails. Is this subset scheduable or not?

Reply:
Scheduler will try to reschedule Pod when scheduled failure by preempting lower priority pods if possible.. In our design, controller will tag this subset unschulable when the subset has one or more than one failed Pod when failed time more than a giving duration.

Reply:
#### How to define a subset is unschedulable
The subset has one or more than one Pod by checking it's status `condition.type=PodScheduled`, 'condition.status=False', and `condition.reason=Unschedulable`

#### Why is it designed
The subset nodes has no available resources to receive all Pods, and it is possible that the failure pods is one or more than one.

Reply:
1. When a pod in a subset has condition[PodScheduled] = false, the subset is considered temporarily unschedulable. The condition[PodScheduled] = false, means that the cluster does not have enough resources to schedule at this moment, even if only one Pod scheduling fails.
2. This state is temporary, and after a period of time, the spread controller will mark it schedulable.